src/core/qobjectuniqueptr.h

```diff
@@ -248,4 +251,263 @@ QObjectUniquePtrFromVariant( const QVariant &variant )
+      {
+        QObject::disconnect( mParentDestroyedConnection );
+      }
+      mParentDestroyedConnection = QObject::connect( parent, &QObject::destroyed, parent, [ = ]()
+      {
+        mParent = nullptr;
+        // parent is being deleted BEFORE child, so it is responsible for deleting the child -- we don't need to delete it here!
+        mChild = nullptr;
+      } );
+      mParent = parent;
+    }
+
+    /**
+     * Assigns a new \a child to the pointer.
+     *
+     * The existing child will be deleted.
+     */
+    inline QObjectParentUniquePtr<T> &operator=( T *child )
+    {
+      if ( child && !mParent )
+      {
+        qWarning() << "Assigning pointer to QObjectParentUniquePtr with nullptr parent.";
+      }
+      delete mChild;
+      mChild = child;
+      return *this;
+    }
+
+    /**
+     * Returns the raw pointer to the managed QObject.
+     */
+    inline T *data() const
+    {
+      return static_cast<T *>( mChild );
+    }
+
+    /**
+     * Returns the raw pointer to the managed child.
+     */
+    inline T *get() const
+    {
+      return static_cast<T *>( mChild );
+    }
+
+    /**
+     * Returns a raw pointer to the managed child.
+     */
+    inline T *operator->() const
+    {
+      return data();
+    }
+
+    /**
+     * Dereferences the managed child.
+     *
+     * \warning Will deference a NULLPTR if the parent has been deleted.
+     */
+    inline T &operator*() const
+    {
+      return *data();
+    }
+
+    /**
+     * Const getter for the managed raw pointer.
+     */
+    inline operator T *() const
+    {
+      return data();
+    }
+
+    /**
+     * Checks if the managed pointer is NULLPTR.
+     */
+    inline bool isNull() const
+    {
+      return !mChild;
+    }
+
+    /**
+     * Checks if the pointer managed by this object is NULLPTR.
+     *
+     * If it is not NULLPTR TRUE will be returned, if it is NULLPTR
+     * FALSE will be returned.
+     */
+    inline operator bool() const
+    {
+      return static_cast< bool >( mChild );
+    }
+
+    /**
+     * Clears the pointer. The managed object is set to NULLPTR and will not be deleted.
+     */
+    inline void clear()
+    {
+      mChild = nullptr;
+    }
+
+    /**
+     * Clears the pointer and returns it.
+     *
+     * The managed object will not be deleted and it is the callers
+     * responsibility to guarantee that no memory is leaked.
+     */
+    inline T *release()
+    {
+      T *p = qobject_cast<T *>( mChild );
+      mChild = nullptr;
+      return p;
+    }
+
+    /**
+     * Will reset the managed pointer to \a p.
+     *
+     * If there is already an object managed currently it will be deleted.
+     *
+     * If \a p is not specified the managed object will be deleted and
+     * this object reset to NULLPTR.
+     */
+    void reset( T *p = nullptr )
+    {
+      if ( p && !mParent )
+      {
+        qWarning() << "Assigning pointer to QObjectParentUniquePtr with nullptr parent.";
```

Comment:
You meant to use qWarning and not QgsDebugMsg here? Just checking.

Reply:
Yep -- it won't show unless something is very wrong, and I want to extra "in your face"-ness which qWarning gives us